[WIP] Added more constants

DavidHusicka opened the pull request:
Added more constants

Used JS as reference of missing mathematical constants.

Added constants are:
E, LN2, LN10, LOG2E, LOG10E, SQRT12, SQRT2

## Why?

These constants will be useful for wide range of computations. Also I noticed that some constants are defined but not accessible from GDScript (this makes them all accessible).

I don't expect this to be in 3.2 as all new feature additions are frozen. Let this be part of 4.0.

DavidHusicka committed:
Added more constants

Added more constats

Used JS as reference of missing mathematical constats.

Added constats are:
E, LN2, LN10, LOG2E, LOG10E, SQRT12, SQRT2

Chaosus commented:
Note, that all of these new constants need to be implemented across the engine including GDNative, VisualScript, and C#, otherwise it is not appropriate for our quality control. Don't hurry, there a lot of time before 4.0 is landed to freeze state.

bydavenet commented:
> across the engine
Ok, I'll have a look.

aaronfranke commented:
It would be nice if you could demonstrate that these constants are useful for games. Features with use cases are better than features without use cases.

bydavenet commented:
Ok, I'll use demonstrate the usage when I make the implementation across the whole engine (I don't have much time lately so it might be pretty late).

DavidHusicka committed:
Full support to engine

New constants are fully supported by GDScript, Editor, Visual Scripts, Visual Script editor, C#

aaronfranke commented on modules/mono/glue/Managed/Files/Mathf.cs:
```diff
@@ -15,6 +15,14 @@ public static partial class Mathf
+        public const real_t Log10e = (real_t) 0.4342944819032518276511289189166050822943970058036M;
+        public const real_t Sqrt12 = (real_t) 0.7071067811865475244008443621048490M;
+        public const real_t Sqrt2 = (real_t) 1.4142135623730950488016887242M;
+        
```
You have an extra new line here. More importantly, you need to remove these from `MathfEx.cs` if you add them here.

aaronfranke reviewed (commented):
Overall, I think E and Sqrt2 are useful, they're also already in C#. Sqrt12 is a variant of Sqrt2 and can be achieved with Sqrt2 / 2 anyway so IMO it's only worth including if it's demonstrated to be very useful.

No clue how useful the log constants are, actually I've never seen Log2e/Log10e used anywhere. These need justification.

While you're at it, we may wish to expose Epsilon. This constant is 1e-6 but becomes 1e-14 if `REAL_T_IS_DOUBLE`.

bydavenet commented:
> While you're at it, we may wish to expose Epsilon. This constant is 1e-6 but becomes 1e-14 if `REAL_T_IS_DOUBLE`.

Ok, I can add Epsilon but I don't know how to make type checking for constants. You can add this constant to this if you know how. Otherwise I would add two new constants (EPSILON, EPSILON_DOUBLE).

DavidHusicka committed:
Resolved issue with Mathf.cs and MathfEx.cs

Calinou commented:
@bydavenet You can probably define the constant conditionally with `#ifdef REAL_T_IS_DOUBLE`.

aaronfranke commented:
@DavidHusicka  Is this still desired? If so, it needs to be rebased on the latest master branch.

DavidHusicka commented:
@aaronfranke I am planning to reimplement this again in a new branch and use this as a little reference since I heard there were huge changes in GDScript in the meantime and it might be better to implement this feature from scratch I guess. It is still desired but you can close this PR since there will come new one using the new branch (I just don't know when I will implement this).
